week03/assignment03.go

```diff
@@ -0,0 +1,79 @@
+package main
+
+import (
+	"fmt"
+)
+
+type Movie struct {
+	Name    string
+	Length  int
+	rating  float32
+	plays   int
+	viewers int
+}
+
+type CritiqueFn func(m *Movie) (float32, error)
+
+type Theater struct{}
+
+func (m *Movie) Rate(rating float32) error {
+
+	if m.plays == 0 {
+		return fmt.Errorf("can't review a movie without watching it first")
+	}
+
+	m.rating = m.rating + rating
+	return nil
+
+}
+
+func (m *Movie) Play(viewers int) {
+	m.viewers = m.viewers + viewers
+	m.plays = m.plays + viewers
+}
+
+func (m Movie) Viewers() int {
+	return m.viewers
+}
+
+func (m Movie) Plays() int {
+	return m.plays
+}
+
+func (m Movie) Rating() float64 {
+	ratings := m.rating / float32(m.plays)
+	return float64(ratings)
+}
+
+func (m Movie) String() string {
+	s := fmt.Sprintf("%v (%dm) %.1f%%", m.Name, m.Length, m.rating)
+	return s
+}
+
+func (t *Theater) Play(viewers int, m ...*Movie) error {
+
+	for _, v := range m {
+		if v.Name == "" && v.Length == 0 {
+			return fmt.Errorf("no movies to play")
+		}
+		v.Play(viewers)
+	}
+
+	return nil
+}
+
+func (t *Theater) Critique(fn CritiqueFn, m []*Movie) error {
+	for _, v := range m {
+		v.Play(1)
+		if rate, err := fn(v); err != nil {
+			return fmt.Errorf("no rate provided or exist")
+		} else {
+			if err := v.Rate(rate); err != nil {
+				return fmt.Errorf("oh no, something went wrong! %v", err)
+			}
+		}
```

Comment:
This set of clauses is far too complex. The logic is hard to follow. In Go, we try to eliminate the need for complex statements like these by trying to not use the `else` clause. 

```go
func (t *Theatre) Critique(fn CritiqueFn, m []*Movie) error {
	for _, v := range m {
		v.Play(1)

		rate, err := fn(v)
		if err != nil {
			return fmt.Errorf("no rate provided or exist")
		}

		if err := v.Rate(rate); err != nil {
			return fmt.Errorf("oh no, something went wrong! %w", err)
		}
	}

	return nil
}
```

Reply:
cool, it's the PHP in me. I will stay away else clauses. I have refactored the code.